### ISSUE TITLE
How to find a child UIItem given an UIItem (as opposed to a window)?

### ISSUE THREAD

**Issue by idvorkin** [TestStack/White#381]
_I'm new to White, love the idea, can't wait to use it - but suspect I may be doing something silly so please correct me_

I have a complex tree, and need to click on a button on a subelement. I can find the parent element, but don't grock the right way to find the child element.  I'll share what I've tried in case it helps answer the question, or makes clear my gaps in understanding White :) 

I have a tree that looks like the below, and I need to click on Button2

      + Complex Stuff
      - List
         + List Item 1: Custom Control
         - List Item 2: Custom Control 
               - Button1 
               + Complex Stuff
               - Button2
        + Complex Stuff

Finding the parent element "just works out of the box": 
     
     var list = window.Get<ListBox>(searchCriteria)
     var parent = list.Items[2];

__Now, I don't know how to find button2 on the parent.__ I'd imagine I could just do: 
  
    var searchCriteria = SearchCriteria.ByClassName().AndByIndex(2);
    var button = parent.Get<Button>(searchCriteria);

However, that doesn't work because parent is a UIItem not a Window :(, so I tried using parent.GetElement: 

    var searchCriteria = SearchCriteria.ByClassName().AndByIndex();
    var button = parent.GetElement(searchCriteria);
    var realButton = new Button(button, window);

Which kind of works, in that it returns the first button, but ignores the ByIndex parameter. From which I conclude I'm doing something wrong :) 

Looking forward to learning what I'm doing wrong :)

**Comment by dpisanu** [comment]
Welcome and thank you for your question.
It looks like you tried a few things already. So let's see how we can get you on the right path.

TestStack.White.UIItems.ListBoxItems.ListBox 
inherrits an Items Property from 
TestStack.White.UIItems.ListBoxItems.ListControl

Items returns 
TestStack.White.UIItems.ListBoxItems.ListItems
through which you can iterate.

So, if you know you ALWAYS want the second ListItem, then you can just use it.

```javascript
var secondListItem = list.Items[1];
```

I would not guarantee that the ListItems is in a right order, so I would use some identification markers, such as AutomationId or Name to identify if the ListItem is the one you want.
You can use LINQ to get the one you want.

Now you have a choice. Fetch All Items inside the ListItem, check some markers and cast the UIItem to a Button or do a Get.
You already tried GET but it obviously did not work.

The ListItem derrives from UIItem.
For UIItem there are extention Methods that allow you to find contained UI Items.
Have a look at the static class TestStack.White.UIItems.WPFUIItems.WPFUIItem.cs

```javascript
var button = secondListItem.Get<Button>(SearchCriteria.ByAutomationId("AutomationIdOfTheSecondButton"));
```

This should give you the second Button.
If your buttons do not have AutomationIDs, set some.
If they have the same for some reason, check if they have other differentiating identifiers, such as Name or Framework.

Hope this helped you.

**Comment by idvorkin** [comment]
Calling get (which I suspect honors byIndex) would be perfect!  

But it doesn't exist on the list item for me.  I think because it's a UIItem, not a window?  Am I missing something?

**Comment by dpisanu** [comment]
Not sure what version you are working on.
If you take the latest Master ->
ListItem derrives from UIItem.
UIItem implements IUIITem Interface.
There are Extention methods for IUIItem which bring you the functionality to call .Get

This is how it is done for a Window as well. Window is also derriving from UIItem.

**Comment by idvorkin** [comment]
__BINGO!  User Error :)__

The problem was I wasn't including the namespace, so couldn't see the extension methods:

      using TestStack.White.UIItems.WPFUIItems; 

With that namespace, everything works as expected.

Huge Thanks!

**Comment by dpisanu** [comment]
Heh. Indeed. You'll need the namespace. 
The extention methods aren't in the same one as the IUIItem ... should have mentioned that explicitely. 

Enjoy your new found powers :)

**Comment by dpisanu** [comment]
You re-opened again. Running into some more problems?

**Comment by idvorkin** [comment]
_Oops- got distracted :(_

Is there extension methods for Win32 UI Items?  

For example, I'm automating Outlook, and I need to apply the same pattern of find the parent, then enumerate it's children.  White tells me the outlook control I'm enumerating is not WPF, so I shouldn't use the WPFUIItem extension method. 

      [Warn - 6:40:57 AM] 'TestStack.White.UIItems.WPFUIItems.WPFUIItem' Only WPF items should be treated as container items

When I use the WPFUIItem's Get extension method it fails to find the item, but GetElement does work: 

Here's a sample if you have Outlook open on the mail pane, and you use grouping,  and want to repro 
```csharp
var workSession = new WorkSession(workConfig, new NullWorkEnvironment());

var app = TestStack.White.Application.Attach("outlook");
var w = app.GetWindow(SearchCriteria.ByClassName("rctrl_renwnd32"), InitializeOption.NoCache);

var mailList = w.Get(SearchCriteria.ByClassName("SuperGrid"));
var firstHeader = mailList.Get<GroupBox>(SearchCriteria.ByClassName("GroupHeader"));
```
Now this code returns an exception: 
```csharp
// fails to find item.
var automationMail = firstHeader.Get(SearchCriteria.ByClassName("ThreadItem"));
```
As a workaround, this alternate code can find the child item, but it suffer from the original bug I described where AndIndex is ignored 

```csharp
var automationMail = firstHeader.GetElement(SearchCriteria.ByClassName("ThreadItem"));
var mail = new CustomUIItem(automationMail, w);
```
As an FYI, here's the partial exception details that White returns with the Get call (which makes me wonder if there's a mix of WPF and Win32 controls??): 
```
Multiple TestControls found for ControlType=tree view item and FrameworkId: - 

TestStack.White.UIItems.TreeItems.Win32TreeNode, 
TestStack.White.UIItems.TreeItems.WPFTreeNode, 
TestStack.White.UIItems.TreeItems.Win32TreeNode

```

**Comment by dpisanu** [comment]
You are correct in your assumption that the Extention Methods, as the Warning states, are only available for WPF Controls. All other Framework Types do not support the Methodology behind the .Get Extentions.
I am not sure if we can Extend these Functions to support the other Frameworks other then WPF.
It does seem to work for Outlook which makes me question if Microsoft is combining different Famework Controls in Outlook. Not completely out of the question.

**Comment by idvorkin** [comment]
_First thanks for your help, I'd be dead in the water with out it :) _

A few observations to raise in case they haven't popped:

1) Get() only partially works for Outlook, and I suspect what's going on is some controls are WPF and some are not, so it works when it's enumerating a WPF control. 
2) If SearchCritieria.ByIndex worked on GetChildElement, users could work around the issue.  
3) I don't know which set of SearchConditions work with GetChild Element.

**Comment by dpisanu** [comment]
The moment you start using IUIITem.GetElement you are working with an System.Windows.Automation.AutomationElement wich is the raw UIA Object.

I have never much bothered digging down to that level. 
It's abit messy and you need to fiddle around alot. Can't be much help in that. Sorry.

Not sure where GetChildElement comes from. Might e a UIA Function. (?)
What you can try though, on the AutomationElement, is calling a specific function.

public AutomationElementCollection FindAll(TreeScope scope, Condition condition);

See if you can get the complete Tree Structure and for the first try iterate through all the items.

**Comment by idvorkin** [comment]
_I'll need more time to dig into this, but stashing my notes here in case it's helpful for others_

I suspect GetChildElement is pretty close to working, with the issue being the code doesn't handle the IndexCondition.   

GetElement already handles taking in the search criteria:
```csharp
public virtual AutomationElement GetElement(SearchCriteria searchCriteria)
{
       return
           new AutomationElementFinder(automationElement).FindDescendantRaw(
              searchCriteria.AutomationSearchCondition);
}
```
The rub, is the ByIndex SearchCiteria, doesn't work by AutomationSearchCondition, but instead by filter:
```csharp
public class IndexCondition : SearchCondition
{
   // <snip>
    public override Condition AutomationCondition
    {
        get { return null; }
    }
    // <snip>
    public virtual IUIItem Filter(UIItemCollection collection)
    {
        var list = new List<IUIItem>(collection);
        list.Sort(new UIItemPositionComparer());
        return list.Count < (index + 1) ? null : list[index];
     }
```
I suspect the key to this puzzle is finding the right place to call ```searchCriteria.Filter()```,

**Comment by vivek9591** [comment]
Try to use this code...
```
public void IteminSubMenuClick(string itemInSubmenu, string uiControl, bool selectLast = false)
{
        if (selectLast)
        {
               TestStack.White.UIItems.IUIItem listViewItem = null;

                   TestStack.White.UIItems.IUIItem[] subMenuElements = mainWindow.GetMultiple(SearchCriteria.ByAutomationId(uiControl));

                    TestStack.White.UIItems.IUIItem activeSubMenu = null;
                    foreach (TestStack.White.UIItems.IUIItem subMenu in subMenuElements)
                    {
                        if (subMenu.Visible)
                        {
                            if (null != activeSubMenu)
                            {
                               //Error if multiple item is there with same name
                            }
                            activeSubMenu = subMenu;
                        }
                    }
                    if (null == activeSubMenu)
                    {
                        //Error if no item is there
                    }
                    listViewItem = TestStack.White.UIItems.WPFUIItems.WPFUIItem.Get(activeSubMenu, SearchCriteria.ByText(itemInSubmenu));
                    if (listViewItem != null)
                    {
                        listViewItem.Click();
                    }
                }            
        }
```
